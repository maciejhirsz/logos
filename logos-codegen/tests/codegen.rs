use std::{error::Error, io, path::PathBuf};

#[rstest::rstest]
#[case("simple")]
#[case("no_error_lut")]
pub fn test_codegen(#[case] fixture: &str) -> Result<(), Box<dyn Error>> {
    let mut fixture_dir = PathBuf::new();
    fixture_dir.push(env!("CARGO_MANIFEST_DIR"));
    fixture_dir.push("tests");
    fixture_dir.push("data");
    fixture_dir.push(fixture);

    let input = fixture_dir.join("input.rs");
    fixture_dir.push("output.rs");
    let output_file = fixture_dir;

    let input = std::fs::read_to_string(input)?;
    let output = std::fs::read_to_string(&output_file)?;

    let generated = logos_codegen::generate(input.parse()?);
    let generated = generated.to_string();

    if std::env::var("BLESS_CODEGEN").is_ok_and(|value| value == "1") {
        std::fs::write(&output_file, &generated)?;
        return Ok(());
    }

    assert_eq!(generated, output, "Codegen test failed: `{fixture}`, run tests again with env var `BLESS_CODEGEN=1` to bless these changes");

    Ok(())
}
