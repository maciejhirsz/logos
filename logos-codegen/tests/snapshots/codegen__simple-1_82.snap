---
source: logos-codegen/tests/codegen.rs
expression: generated
---
impl < 's > :: logos :: Logos < 's > for Token { type Error = () ; type Extras = () ; type Source = str ; fn lex (lex : & mut :: logos :: Lexer < 's , Self >) -> Option < Result < Self , Self :: Error >> { use :: logos :: internal :: { LexerInternal , CallbackRetVal , CallbackResult , SkipRetVal , SkipResult , } ; use :: logos :: Logos ; type Lexer < 's > = :: logos :: Lexer < 's , Token > ; # [derive (Clone , Copy)] enum LogosState { State1024Ctx0 , State1536 , State2048 , State2560 , State3072 } const START : LogosState = LogosState :: State1536 ; let mut state = START ; let mut offset = lex . offset () ; loop { match state { LogosState :: State1024Ctx0 => { lex . end (offset - 1) ; match lex . read :: < u8 > (offset) { _ => { return Some (Ok (Token :: Letter)) ; } } } LogosState :: State1536 => { match lex . read :: < u8 > (offset) { Some (b'a') => { offset += 1 ; state = LogosState :: State3072 ; } , None => return None , _ => { lex . end_to_boundary (offset + 1) ; return Some (Err (Self :: Error :: default ())) ; } } } LogosState :: State2048 => { match lex . read :: < u8 > (offset) { Some (b'z') => { offset += 1 ; state = LogosState :: State2560 ; } , _ => { lex . end_to_boundary (offset + 1) ; return Some (Err (Self :: Error :: default ())) ; } } } LogosState :: State2560 => { match lex . read :: < u8 > (offset) { Some (0u8 ..= 255u8) => { offset += 1 ; state = LogosState :: State1024Ctx0 ; } , None => { offset += 1 ; state = LogosState :: State1024Ctx0 ; } _ => { lex . end_to_boundary (offset + 1) ; return Some (Err (Self :: Error :: default ())) ; } } } LogosState :: State3072 => { match lex . read :: < u8 > (offset) { Some (b'-') => { offset += 1 ; state = LogosState :: State2048 ; } , _ => { lex . end_to_boundary (offset + 1) ; return Some (Err (Self :: Error :: default ())) ; } } } } } } }
